### PR TITLE
curvefs/tools-vs&client: warmupPath

### PR DESCRIPTION
**Pull request by Cyber-SiKu**
Signed-off-by: Cyber-SiKu <Cyber-SiKu@outlook.com>

### What problem does this PR solve?

Issue Number: #xxx 

Problem Summary:

### What is changed and how it works?

What's Changed:

How it Works:

Side effects(Breaking backward compatibility? Performance regression?):

### Check List

- [ ] Relevant documentation/comments is changed or added
- [ ] I acknowledge that all my contributions will be made under the project's license

### CODE REVIEW

File: curvefs/src/client/curve_fuse_op.cpp [hunk 1]
```diff
@@ -273,6 +276,60 @@ void FuseOpGetAttr(fuse_req_t req, fuse_ino_t ino, struct fuse_file_info *fi) {
...
+    int ret = 0;
+    switch (curvefs::client::common::GetWarmupType(type)) {
+        case curvefs::client::common::WarmupType::WarmupTypeList:
+            g_ClientInstance->PutWarmTask(path);
+            break;
+        case curvefs::client::common::WarmupType::WarmupTypeSingle:
+            g_ClientInstance->FetchDentryEnqueue(path);
+            break;
+        default:
+            ret = ERANGE;
```
Comment (wu-hanqing) [inline curvefs/src/client/curve_fuse_op.cpp:289]: what's the range means?

File: curvefs/src/client/common/common.h [hunk 2]
```diff
@@ -64,6 +65,26 @@ std::ostream &operator<<(std::ostream &os, MetaServerOpType optype);
 
 const uint32_t MAXXATTRLENGTH = 256;
 
+const char kCurveFsWarmupXAttr[] = "curvefs.warmup.op";
+const char kCurveFsWarmupOpAdd[] = "add";
+const char kCurveFsWarmupTypeList[] = "list";
+const char kCurveFsWarmupTypeSingle[] = "single";
+
+enum WarmupOpType {
```
Comment (wu-hanqing) [inline curvefs/src/client/common/common.h:73]: `enum class`?

File: curvefs/src/client/common/common.h [hunk 3]
```diff
@@ -64,6 +65,26 @@ std::ostream &operator<<(std::ostream &os, MetaServerOpType optype);
...
+    WarmupOpUnknown = 0,
+    WarmupOpAdd = 1,
```
Comment (wu-hanqing) [inline curvefs/src/client/common/common.h:75]: add prefix `k`

File: curvefs/src/client/common/common.h [hunk 4]
```diff
@@ -64,6 +65,26 @@ std::ostream &operator<<(std::ostream &os, MetaServerOpType optype);
...
+};
+
+WarmupOpType GetWarmupOpType(const std::string& op);
+
+enum WarmupType {
+    WarmupTypeUnknown = 0,
+    WarmupTypeList = 1,
+    WarmupTypeSingle = 2,
+};
```
Comment (wu-hanqing) [inline curvefs/src/client/common/common.h:84]: ditto

File: curvefs/src/client/common/common.h [hunk 5]
```diff
@@ -64,6 +65,26 @@ std::ostream &operator<<(std::ostream &os, MetaServerOpType optype);
 
 const uint32_t MAXXATTRLENGTH = 256;
 
+const char kCurveFsWarmupXAttr[] = "curvefs.warmup.op";
+const char kCurveFsWarmupOpAdd[] = "add";
+const char kCurveFsWarmupTypeList[] = "list";
+const char kCurveFsWarmupTypeSingle[] = "single";
```
Comment (wu-hanqing) [inline curvefs/src/client/common/common.h:71]: is there anywhere uses them apart from `common.cpp`, if so, move them to cpp file.

File: curvefs/src/client/common/common.h [hunk 6]
```diff
@@ -64,6 +65,26 @@ std::ostream &operator<<(std::ostream &os, MetaServerOpType optype);
 
 const uint32_t MAXXATTRLENGTH = 256;
 
+const char kCurveFsWarmupXAttr[] = "curvefs.warmup.op";
+const char kCurveFsWarmupOpAdd[] = "add";
```
Comment (wu-hanqing) [inline curvefs/src/client/common/common.h:69]: should you add prefix `curvefs.warmup.add`?

File: curvefs/src/client/curve_fuse_op.cpp [hunk 7]
```diff
@@ -273,6 +276,60 @@ void FuseOpGetAttr(fuse_req_t req, fuse_ino_t ino, struct fuse_file_info *fi) {
     fuse_reply_attr(req, &attr, g_fuseClientOption->attrTimeOut);
 }
 
+int addWarmup(const std::string&& type, const std::string& path) {
```
Comment (wu-hanqing) [inline curvefs/src/client/curve_fuse_op.cpp:279]: CamelCase, and why first arguemtn is a const rvalue reference?

File: curvefs/src/client/curve_fuse_op.cpp (hunk 1 above)
Comment (wu-hanqing) [inline curvefs/src/client/curve_fuse_op.cpp:289]: what's the meaning of `ERANGE`?

File: curvefs/src/client/curve_fuse_op.cpp [hunk 8]
```diff
@@ -273,6 +276,60 @@ void FuseOpGetAttr(fuse_req_t req, fuse_ino_t ino, struct fuse_file_info *fi) {
...
+        case curvefs::client::common::WarmupType::WarmupTypeSingle:
+            g_ClientInstance->FetchDentryEnqueue(path);
+            break;
+        default:
+            ret = ERANGE;
+    }
+    return ret;
+}
+
+int warmup(const std::string& name, const std::string& value) {
```
Comment (wu-hanqing) [inline curvefs/src/client/curve_fuse_op.cpp:294]: ditto

File: curvefs/src/client/curve_fuse_op.cpp (hunk 1 above)
Comment (Cyber-SiKu) [reply curvefs/src/client/curve_fuse_op.cpp:289]: > what's the range means?

fix

File: curvefs/src/client/common/common.h (hunk 2 above)
Comment (Cyber-SiKu) [reply curvefs/src/client/common/common.h:73]: > `enum class`?

fix

File: curvefs/src/client/common/common.h (hunk 5 above)
Comment (Cyber-SiKu) [reply curvefs/src/client/common/common.h:71]: > is there anywhere uses them apart from `common.cpp`, if so, move them to cpp file.

just kCurveFsWarmupXAttr will be used by [FuseOpSetXattr](https://github.com/Cyber-SiKu/curve/blob/curvefs/toolv2_client_warmup/curvefs/src/client/curve_fuse_op.cpp)

File: curvefs/src/client/common/common.h (hunk 6 above)
Comment (Cyber-SiKu) [reply curvefs/src/client/common/common.h:69]: > should you add prefix `curvefs.warmup.add`?

no，add means warmup operation，will add del/query in feature

File: curvefs/src/client/curve_fuse_op.cpp (hunk 7 above)
Comment (Cyber-SiKu) [reply curvefs/src/client/curve_fuse_op.cpp:279]: > CamelCase, and why first arguemtn is a const rvalue reference?

bug，I wrote  PutWarmTask and FetchDentryEnqueue  before, used rvalues.

File: curvefs/src/client/curve_fuse_op.cpp (hunk 1 above)
Comment (Cyber-SiKu) [reply curvefs/src/client/curve_fuse_op.cpp:289]: > what's the meaning of `ERANGE`?

xatter name or value

**Review by wuhongsong — APPROVED** [review]

**Review by wu-hanqing — APPROVED** [review]

**Pull request merged**